### DISCUSSION TITLE
Unable to copy text from lazyvim running in remote ssh host to host clipboard

### DISCUSSION THREAD

**Discussion by rbhanot4739** [LazyVim/LazyVim#4602] [Q&A]
Hello,

So I SSHing to my linux host from my mac host machine and have LazyVim running on the ssh host. When I copy anything from within LazyVim, the yanked text is not copied to the mac system clipboard. 

However outside of the neovim if I copy anything using tmux copy mode, the text is correctly copied to the system clipboard. I did check the default LazyVim config [here](https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/config/options.lua#L65C43-L65C56) and was expecting this to work but maybe either my understanding of this incorrect.

```
-- only set clipboard if not in ssh, to make sure the OSC 52
-- integration works automatically. Requires Neovim >= 0.10.0
opt.clipboard = vim.env.SSH_TTY and "" or "unnamedplus" -- Sync with system clipboard
```

So can someone please help here ?

**Comment by dotfrag** [comment]
I'm having the same problem.. I tried to figure it out several times but no luck.. I don't know what the issue is.

Would really appreciate it if anyone has solved this.

FWIW, OSC52 is working, if I run `printf $'\e]52;c;%s\a' "$(base64 <<<'hello world')"`, my clipboard contains `hello world`.

I have tried kitty and wezterm, and also forcing osc52 as documented in `clipboard-osc52`.

**Reply by dpetka2001** [reply]
Can you try setting `:set clipboard=unnamedplus` while in ssh session?

**Reply by dotfrag** [reply]
Nothing changes.

**Comment by rbhanot4739** [comment]
@dpetka2001 I did try that before posting this unfortunately it didn't work

**Reply by dpetka2001** [reply]
What does `:set clipboard?` show?

**Reply by dpetka2001** [reply]
Are you using tmux?

**Reply by rbhanot4739** [reply]
`set clipboard` shows `unnmedplus` and yes I am using tmux however this issue happens even outside of tmux, i.e. if i ssh directly to host outside of a tmux session on the host.

Also if I start a tmux session on the ssh host and copy something with tmux copy mode the text is correctly copied to the host system(mac in my case) clipbord.

**Reply by dpetka2001** [reply]
I don't have an extra PC to ssh into, so I'm just sshing into localhost without tmux and when I set `:set clipboard=unnamedplus` I can copy to host clipboard. I've read that tmux requires additional setup for this to work, but haven't really used it. That's why I only ssh without using tmux.

**Reply by dpetka2001** [reply]
Can you try what is mentioned in this [comment](https://github.com/neovim/neovim/issues/28611#issuecomment-2147744670)? Put the code snippet in your `options.lua` file.

**Reply by dotfrag** [reply]
I have added the following in my options.lua:

```lua
-- Clipboard
if vim.env.SSH_TTY then
  local function my_paste(reg)
    return function()
      local content = vim.fn.getreg('"')
      return vim.split(content, "\n")
    end
  end

  opt.clipboard:append("unnamedplus")
  vim.g.clipboard = {
    name = "OSC 52",
    copy = {
      ["+"] = require("vim.ui.clipboard.osc52").copy("+"),
      ["*"] = require("vim.ui.clipboard.osc52").copy("*"),
    },
    paste = {
      ["+"] = my_paste("+"),
      ["*"] = my_paste("*"),
    },
  }
end
```

And it's working now! I have no idea why but it works. As the next reply in the issue says, yanky stops working, but thankfully I don't care about yanky in ssh sessions.

I have also added the following:

```lua
return {
  "gbprod/yanky.nvim",
  enabled = not vim.env.SSH_TTY,
}
```

I'm doing some more testing.. I'll post back

**Reply by dpetka2001** [reply]
I also ended up disabling yanky because it was causing lag between copying and was getting a warning message about terminal time out.

**Reply by dotfrag** [reply]
Ok, so.. I kept yanky disabled, BUT, the snippet from the linked issue is not really necessary.

This will NOT work:

```
opt.clipboard="unnamedplus"
```

But this works:

```lua
opt.clipboard:append("unnamedplus")
```

I still don't understand why. This also goes against what the docs say:

> Nvim will attempt to automatically determine if the host terminal emulator
supports the OSC 52 sequence and enable the OSC 52 clipboard provider if it
does as long as all of the following are true:


>   • Nvim is running in the |TUI|
  • |g:clipboard| is unset
  • 'clipboard' is not set to "unnamed" or "unnamedplus"
  • $SSH_TTY is set

> If any of the above conditions are not met then the OSC 52 clipboard provider
will not be used by default and Nvim will fall back to discovering a
|clipboard-tool| through the usual process.

**Reply by dotfrag** [reply]
Also LazyVim does some initial setup with clipboard.. I don't have the mental capacity to go over it now, but maybe that causes some issue.

**Reply by dpetka2001** [reply]
@dotfrag Can you change line 293 in `lua/config/init.lua` into `lazy_clipboard = vim.env.SSH_TTY and nil or vim.opt.clipboard` and see if that makes a difference?

**Reply by dotfrag** [reply]
Nope.. ` set clipboard?   clipboard=`

I'm on LazyVim starter.

**Reply by dpetka2001** [reply]
Then I don't think what LazyVim does with clipboard really affects that, because LazyVim essentially just defers setting the value of `clipboard` until on `VeryLazy`. Something else must be going on. You could also check with `nvim --clean` so that you get vanilla Neovim configuration where `clipboard` isn't set and the requirements should be set for osc-52 to work out of the box.

**Reply by dotfrag** [reply]
Still doesn't work :(

If I run this afterwards, then it works:

```
:lua vim.opt.clipboard:append("unnamedplus")
```

or 

```
:set clipboard+=unnamedplus
```

**Reply by dpetka2001** [reply]
Did you try with `nvim --clean`?

**Reply by dotfrag** [reply]
Yes, the above is with `nvim --clean`.

**Reply by dpetka2001** [reply]
Ok I found another [comment](https://github.com/neovim/neovim/issues/29504#issuecomment-2226374704) which basically says

> So do we really need to make sure clipboard is not set to unnamedplus as mentioned in the docs?

> The reason we don't set it automatically when 'clipboard' is set to a non-default value is because copying via OSC 52 can be slow, and pasting can be even slower or (even worse) display an annoying prompt to the user every time they use p. If users want to use OSC 52 with 'unnamedplus' anyway then they can configure OSC 52 manually, but we don't want to make that the default experience.

So, I guess they leave it up to the users to set it manually in their personal configuration and Neovim doesn't set it itself automatically.

As to why you have to use `append`, I really have no idea.

PS: At least it seems it has nothing to do with LazyVim and it's a Neovim default.

**Reply by dotfrag** [reply]
I don't know if it means anything, but the example in the docs also uses append:

```
To ALWAYS use the clipboard for ALL operations (instead of interacting with
the "+" and/or "*" registers explicitly):
    set clipboard+=unnamedplus

```

**Reply by dotfrag** [reply]
I'm having the problem described with slow copying and prompt on pasting.. this is very annoying, I might as well just disable it all together.

**Reply by dpetka2001** [reply]
I really don't use this feature, so I don't know any more than what I've already mentioned and linked to. Sorry I can't be of more help. Maybe Folke knows something of more substance when he gets back.

The slowness is supposed to be gone when you apply the code snippet above, that you said it's not needed after all.

**Reply by dpetka2001** [reply]
You might also want to take a look at [nvim-osc52](https://github.com/ojroques/nvim-osc52) that was created for using `osc` feature before it was incorporated into Neovim >=0.10. Maybe it doesn't have this slowness? Maybe it does? Maybe it's worth checking it out?

I also found another [post](https://gronskiy.com/posts/2023-03-26-copy-via-vim-tmux-ssh/) about using that and copy/paste over ssh using tmux. You might also want to check that out? Not sure if it has the same drawbacks as native Neovim `osc-52` feature.

**Reply by dotfrag** [reply]
I got a bit lost with all the changes I've tried. I put the snippet back to test. Will probably leave this in my config (https://github.com/cameronr/kickstart-modular.nvim/commit/b567fdb93936201ac7b5c20d590b423dcaad8fbf#diff-ebf84919960f20d62c1c0f8c1a170f81bfb34668c6a9d3190f4679ef7add614eR24):

```lua
-- Clipboard
if vim.env.SSH_TTY then
  opt.clipboard:append("unnamedplus")
  local function paste(_)
    return function()
      local content = vim.fn.getreg('"')
      return vim.split(content, "\n")
    end
  end
  vim.g.clipboard = {
    name = "OSC 52",
    copy = {
      ["+"] = require("vim.ui.clipboard.osc52").copy("+"),
      ["*"] = require("vim.ui.clipboard.osc52").copy("*"),
    },
    paste = {
      ["+"] = paste("+"),
      ["*"] = paste("*"),
    },
  }
end
```

I can basically copy from ssh but not paste, I can use Ctrl+Shift+V for that. I'm happy :)

Thanks for all you help and finding the resources, it's been very helpful.

EDIT: I don't use tmux in this scenario. I have different workflow for that.

EDIT2: I'll give the plugins a go.

EDIT3: Works okay, but pretty much same as native support.

**Comment by dotfrag** [comment]
With help from @dpetka2001, this is my final snippet in `options.lua`:

```lua
-- Clipboard
if vim.env.SSH_TTY then
  opt.clipboard:append("unnamedplus")
  local function paste()
    return vim.split(vim.fn.getreg(""), "\n")
  end
  vim.g.clipboard = {
    name = "OSC 52",
    copy = {
      ["+"] = require("vim.ui.clipboard.osc52").copy("+"),
      ["*"] = require("vim.ui.clipboard.osc52").copy("*"),
    },
    paste = {
      ["+"] = paste,
      ["*"] = paste,
    },
  }
end
```

**Reply by dpetka2001** [reply]
I also saw something in the file you posted in your repo. You have a autocmd for removing `formatoptions` on `BufEnter` event. That will be triggered every time you enter a buffer (a bit too much in my personal opinion). `options.lua` will be loaded at the very start when LazyVim starts. The thing is that `FileType` event will get triggered later. That's why Neovim's default `ftplugin` for languages still overwrites your custom `formatoptions`. If you put your autocmd in `lua/config/autocmds.lua` on event `FileType` you shouldn't have any problems and it will only trigger once when the filetype of the file will be detected. That's how I have it as well in my [dotfiles](https://github.com/dpetka2001/dotfiles/blob/4ae0b9e9a67e2a37a4fee7773a8c876d1ac890f3/dot_config/nvim/lua/config/autocmds.lua#L29-L42).

PS: Hope this is not considered as nitpicking. Apologies if it is.

**Reply by dotfrag** [reply]
Quite the opposite, I have the repo public for a reason :)

I pushed now. Do you mean to just change BufEnter to Filetype? I think it's behaving unexpectedly.

~Hold on, maybe I should have followed your code better.~ Nvm, I'm not sure (:

**Reply by dpetka2001** [reply]
Just put the autocmd in `~/.config/nvim/lua/config/autocmds.lua` instead of `options.lua`. That's what I mean and also use `FileType` instead of `BufEnter`. I have it there and it works as expected for me when I press `o/O`.

**Reply by dotfrag** [reply]
It's already in autocmds no?

**Reply by dpetka2001** [reply]
The link that you posted https://github.com/cameronr/kickstart-modular.nvim/commit/b567fdb93936201ac7b5c20d590b423dcaad8fbf shows the autocmd in `options.lua`. Also, now that I notice that link is not a LazyVim configuration. Maybe you linked wrong repo? My assumption was based on the link you provided.

**Reply by dotfrag** [reply]
Ahh I see the confusion, that commit is a linked commit from the [link you provided earlier](https://github.com/neovim/neovim/issues/28611#issuecomment-2147744670). Not my config :)

My config is https://github.com/dotfrag/nvim and I thought you were talking about this

https://github.com/dotfrag/nvim/blob/6279b3f70812e472707472535da553e4d11b4408/lua/config/autocmds.lua#L48-L55

**Reply by dpetka2001** [reply]
Hahaha yep definitely was confused. I thought that the link you mentioned was the diff of the changes you applied to your config. Please disregard :smile:

**Comment by totoroyyb** [comment]
Wanna second on this issue. I would expect the Lazyvim to work out of the box with osc52. At least my `checkhealth` shows:

```
Clipboard (optional)
- OK Clipboard tool found: OSC 52
```

However, it seems to have issues with OSC52. Run `:set clipboard=unnamedplus` does do the trick. However, this seems to be contradict to nvim's clipboard-osc52 doc, where it says:

> When Nvim is running in the |TUI|, it will automatically attempt to determine if the host terminal emulator supports OSC 52. If it does, then Nvim will use OSC52 for copying and pasting if no other clipboard-tool is found and when 'clipboard' is unset.
 
The doc sounds like OSC52 should automatically work when the clipboard is unset, but in my case, I have to set the clipboard to unnamedplus to make it work.

I am unsure if this is a Neovim or lazyvim issue. FYI, I am using Ghostty, but iTerm2 has the same issue, so this is probably terminal-related.

Some additional info:
NVIM v0.11.0
Build type: Release
LuaJIT 2.1.1741730670
